src/System.Linq/src/System/Linq/Partition.cs

```diff
@@ -389,5 +389,308 @@ public int GetCount(bool onlyIfCheap)
+                        TSource result;
+
+                        do
+                        {
+                            remaining--;
+                            result = en.Current;
+                        }
+                        while (remaining >= comparand && en.MoveNext());
+
+                        found = true;
+                        return result;
+                    }
+                }
+
+                found = false;
+                return default(TSource);
+            }
+
+            public TSource[] ToArray()
+            {
+                using (IEnumerator<TSource> en = _source.GetEnumerator())
+                {
+                    if (SkipBeforeFirst(en) && en.MoveNext())
+                    {
+                        int remaining = Limit - 1; // Max number of items left, not counting the current element.
+                        int comparand = HasLimit ? 0 : int.MinValue; // If we don't have an upper bound, have the comparison always return true.
+                        var builder = new LargeArrayBuilder<TSource>(initialize: true);
+
+                        do
+                        {
+                            remaining--;
+                            builder.Add(en.Current);
+                        }
+                        while (remaining >= comparand && en.MoveNext());
+
+                        return builder.ToArray();
+                    }
+                }
+
+                return Array.Empty<TSource>();
+            }
+
+            public List<TSource> ToList()
+            {
+                var list = new List<TSource>();
+
+                using (IEnumerator<TSource> en = _source.GetEnumerator())
+                {
+                    if (SkipBeforeFirst(en) && en.MoveNext())
+                    {
+                        int remaining = Limit - 1; // Max number of items left, not counting the current element.
+                        int comparand = HasLimit ? 0 : int.MinValue; // If we don't have an upper bound, have the comparison always return true.
+
+                        do
+                        {
+                            remaining--;
+                            list.Add(en.Current);
+                        }
+                        while (remaining >= comparand && en.MoveNext());
+                    }
+                }
+
+                return list;
+            }
+
+            private bool SkipBeforeFirst(IEnumerator<TSource> en) => SkipBefore(_minIndexInclusive, en);
+
+            private static bool SkipBefore(int index, IEnumerator<TSource> en) => SkipAndCount(index, en) == index;
+
+            private static int SkipAndCount(int index, IEnumerator<TSource> en)
+            {
+                Debug.Assert(index >= 0);
+                return (int)SkipAndCount((uint)index, en);
+            }
+
+            private static uint SkipAndCount(uint index, IEnumerator<TSource> en)
```

Comment:
Had to add a `uint` overload of this function to support scenarios where `_maxIndexInclusive` is exactly int.MaxValue & we call `Count()`, for example `e.Skip(1).Take(int.MaxValue).Count()`. We don't lose any validation at callsites other than `GetCount` though, since the signed wrapper asserts.